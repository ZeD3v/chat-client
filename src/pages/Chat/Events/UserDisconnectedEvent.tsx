import React from 'react';
import { format } from 'date-fns';

import { User } from '../../../App';
import { ChatEvents } from '../index';

import styles from './styles/ConnectionStatusEvent.module.scss';

type UserDisconnectedEventPayload = {
  username: string;
  reason: string;
};

type UserDisconnectedEvent = {
  type: ChatEvents.USER_DISCONNECTED;
  id: string;
  timestamp: number;
  payload: UserDisconnectedEventPayload;
};

type UserDisconnectedEventProps = {
  user: User | null;
  event: UserDisconnectedEvent;
};

const messages: { [reason: string]: string } = {
  USER_LEFT: 'left',
  USER_INACTIVE: 'disconnected due to inactivity',
};

const UserDisconnectedEvent: React.FC<UserDisconnectedEventProps> = ({
  user,
  event,
}: UserDisconnectedEventProps) => (
  <div className={styles.container}>
    {user?.username === event.payload.username ? 'You' : event.payload.username}{' '}
    {messages[event.payload.reason] || 'disconnected'} -{' '}
    {format(new Date(event.timestamp), 'dd/MM/yyyy HH:mm:ss')}
  </div>
);

export default UserDisconnectedEvent;
