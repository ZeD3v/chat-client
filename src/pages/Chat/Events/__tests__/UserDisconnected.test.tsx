import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import UserDisconnectedEvent from '../UserDisconnectedEvent';

import { ChatEvents } from '../../index';

describe('UserDisconnectedEvent', () => {
  test('renders correctly if user disconnected due to inactivity', () => {
    const { asFragment } = render(
      <UserDisconnectedEvent
        user={{ username: 'Test user 1', token: 'test-token' }}
        event={{
          id: 'testid-1',
          timestamp: 1589041780,
          type: ChatEvents.USER_DISCONNECTED,
          payload: {
            username: 'Test user 2',
            reason: 'USER_INACTIVE',
          },
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders correctly if user left', () => {
    const { asFragment } = render(
      <UserDisconnectedEvent
        user={{ username: 'Test user 1', token: 'test-token' }}
        event={{
          id: 'testid-1',
          timestamp: 1589041780,
          type: ChatEvents.USER_DISCONNECTED,
          payload: {
            username: 'Test user 2',
            reason: 'USER_LEFT',
          },
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
