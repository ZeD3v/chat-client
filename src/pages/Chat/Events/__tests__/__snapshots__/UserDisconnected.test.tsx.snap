// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDisconnectedEvent renders correctly if user disconnected due to inactivity 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    Test user 2 disconnected due to inactivity - 19/01/1970 09:24:01
  </div>
</DocumentFragment>
`;

exports[`UserDisconnectedEvent renders correctly if user left 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    Test user 2 left - 19/01/1970 09:24:01
  </div>
</DocumentFragment>
`;
